File: tests/python/test_camera.py
#!/usr/bin/env python3
import sys
sys.path.insert(0, 'python')

import numpy as np
import hybrid_sfm as sfm

def test_camera_models():
    """Test different camera models."""
    print("Testing camera models...")
    
    # Test PINHOLE model
    camera_pinhole = sfm.Camera(sfm.CameraModel.PINHOLE, 640, 480)
    print(f"✅ Created PINHOLE camera: {camera_pinhole.get_width()}x{camera_pinhole.get_height()}")
    
    # Test projection/unprojection
    point_3d = np.array([1.0, 2.0, 5.0])
    point_2d = camera_pinhole.world_to_image(point_3d)
    print(f"✅ Projected 3D {point_3d} -> 2D {point_2d}")
    
    point_3d_recovered = camera_pinhole.image_to_world(point_2d, 5.0)
    print(f"✅ Unprojected 2D {point_2d} -> 3D {point_3d_recovered}")
    
    # Check error
    error = np.linalg.norm(point_3d - point_3d_recovered)
    assert error < 1e-6, f"Projection error too large: {error}"
    print("✅ Projection/unprojection cycle accurate")
    
    # Test intrinsic matrix
    K = camera_pinhole.get_K()
    print(f"✅ Intrinsic matrix K:\n{K}")
    
    # Test COLMAP serialization
    colmap_str = camera_pinhole.to_COLMAP()
    print(f"✅ COLMAP format: {colmap_str}")

def test_camera_pose():
    """Test camera pose operations."""
    print("\nTesting camera pose...")
    
    # Create poses
    pose1 = sfm.CameraPose()
    R = np.eye(3)
    R[0,0] = 0; R[0,1] = -1
    R[1,0] = 1; R[1,1] = 0  # 90 degree rotation
    t = np.array([1.0, 2.0, 3.0])
    pose2 = sfm.CameraPose(R, t)
    
    # Test pose composition
    pose3 = pose1 * pose2
    print("✅ Pose composition working")
    
    # Test inverse
    pose_inv = pose2.inverse()
    identity = pose2 * pose_inv
    print("✅ Pose inverse working")
    
    # Test world/camera transforms
    point_world = np.array([1.0, 0.0, 0.0])
    point_camera = pose2.world_to_camera(point_world)
    point_world_recovered = pose2.camera_to_world(point_camera)
    
    error = np.linalg.norm(point_world - point_world_recovered)
    assert error < 1e-6, f"Transform error too large: {error}"
    print("✅ World/camera transforms accurate")

def test_distortion():
    """Test camera distortion models."""
    print("\nTesting distortion models...")
    
    # Create camera with radial distortion
    params = [500.0, 320.0, 240.0, -0.1, 0.05]  # f, cx, cy, k1, k2
    camera_radial = sfm.Camera(sfm.CameraModel.RADIAL, 640, 480)
    camera_radial.set_params(params)
    
    # Test that distortion affects projection
    point_3d = np.array([1.0, 1.0, 10.0])
    
    # Compare with no distortion
    camera_simple = sfm.Camera(sfm.CameraModel.SIMPLE_PINHOLE, 640, 480)
    point_2d_simple = camera_simple.world_to_image(point_3d)
    point_2d_radial = camera_radial.world_to_image(point_3d)
    
    diff = np.linalg.norm(point_2d_simple - point_2d_radial)
    assert diff > 0.1, "Distortion should have effect"
    print(f"✅ Distortion effect: {diff:.2f} pixels")
    
    # Test unprojection accuracy
    point_3d_recovered = camera_radial.image_to_world(point_2d_radial, 10.0)
    error = np.linalg.norm(point_3d - point_3d_recovered)
    assert error < 1e-4, f"Unprojection error too large: {error}"
    print("✅ Distortion unprojection accurate")

if __name__ == "__main__":
    print("Day 2: Camera Models & Calibration Tests\n")
    
    test_camera_models()
    test_camera_pose()
    test_distortion()
    
    print("\n🎉 Day 2 Complete! Camera models and calibration working!")

File: tests/cpp/test_camera.cpp
#include <gtest/gtest.h>
#include "core/types/camera.h"
#include <random>

using namespace hybrid_sfm;

class CameraTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Create test cameras
        simple_pinhole_ = Camera(Camera::Model::SIMPLE_PINHOLE, 640, 480);
        pinhole_ = Camera(Camera::Model::PINHOLE, 640, 480);
        
        // Camera with distortion
        std::vector<double> radial_params = {500.0, 320.0, 240.0, -0.1, 0.05};
        radial_ = Camera(Camera::Model::RADIAL, 640, 480, radial_params);
    }
    
    Camera simple_pinhole_;
    Camera pinhole_;
    Camera radial_;
};

TEST_F(CameraTest, ProjectUnprojectSimplePinhole) {
    Point3D point3d(1.0, 2.0, 5.0);
    Point2D point2d = simple_pinhole_.worldToImage(point3d);
    
    // Check projection
    // For SIMPLE_PINHOLE: f=640, cx=320, cy=240
    // Normalized: (1/5, 2/5) = (0.2, 0.4)
    // Image: (0.2*640 + 320, 0.4*640 + 240) = (448, 496)
    EXPECT_NEAR(point2d.x(), 448.0, 1e-6);
    EXPECT_NEAR(point2d.y(), 496.0, 1e-6);
    
    // Check unprojection
    Point3D unprojected = simple_pinhole_.imageToWorld(point2d, 5.0);
    EXPECT_NEAR(unprojected.x(), point3d.x(), 1e-6);
    EXPECT_NEAR(unprojected.y(), point3d.y(), 1e-6);
    EXPECT_NEAR(unprojected.z(), point3d.z(), 1e-6);
}

TEST_F(CameraTest, DistortionRadial) {
    Point3D point3d(1.0, 2.0, 10.0);
    Point2D point2d = radial_.worldToImage(point3d);
    
    // Create simple camera with same focal length for comparison
    std::vector<double> simple_params = {500.0, 320.0, 240.0};
    Camera simple(Camera::Model::SIMPLE_PINHOLE, 640, 480, simple_params);
    Point2D simple_point = simple.worldToImage(point3d);
    
    // With distortion, the point should be different from simple projection
    EXPECT_GT((point2d - simple_point).norm(), 0.1);
    
    // Unprojection should recover the original point
    Point3D unprojected = radial_.imageToWorld(point2d, 10.0);
    EXPECT_NEAR(unprojected.x(), point3d.x(), 1e-4);
    EXPECT_NEAR(unprojected.y(), point3d.y(), 1e-4);
    EXPECT_NEAR(unprojected.z(), point3d.z(), 1e-4);
}

TEST_F(CameraTest, COLMAPSerialization) {
    std::string colmap_str = pinhole_.toCOLMAP();
    
    // Should contain model name and parameters
    EXPECT_TRUE(colmap_str.find("PINHOLE") != std::string::npos);
    EXPECT_TRUE(colmap_str.find("640") != std::string::npos);
    EXPECT_TRUE(colmap_str.find("480") != std::string::npos);
}

TEST_F(CameraTest, IntrinsicMatrix) {
    Eigen::Matrix3d K = pinhole_.getK();
    
    EXPECT_EQ(K(0, 0), pinhole_.getFocalLengthX());
    EXPECT_EQ(K(1, 1), pinhole_.getFocalLengthY());
    EXPECT_EQ(K(0, 2), pinhole_.getPrincipalPointX());
    EXPECT_EQ(K(1, 2), pinhole_.getPrincipalPointY());
    EXPECT_EQ(K(2, 2), 1.0);
}

TEST_F(CameraTest, ModelTypes) {
    // Test different camera models
    Camera opencv(Camera::Model::OPENCV, 640, 480);
    EXPECT_EQ(opencv.getParams().size(), 8);
    
    Camera fisheye(Camera::Model::OPENCV_FISHEYE, 640, 480);
    EXPECT_EQ(fisheye.getParams().size(), 8);
    
    // Test model name lookup
    Camera::Model model = Camera::modelFromName("RADIAL");
    EXPECT_EQ(model, Camera::Model::RADIAL);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}